flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeNestedField.scala

```diff
@@ -0,0 +1,295 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.utils
+
+import org.apache.flink.table.api.TableException
+
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rex._
+
+import java.util
+import java.util.{List => JList, Map => JMap}
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+/**
+ * RexNodeNestedField is a tree node to build the used fields tree.
+ *
+ * @param name      The name of the fields in the origin schema
+ * @param index     The index of the field in the origin schema. It
+ *                  only works for the RowType.
+ * @param fieldType The type of the field. It is useful when
+ *                  rewriting the projections. It uses the LinkedHashMap
+ *                  to keep the insert order. In some cases, it can
+ *                  reduce the cost of the reorder of the fields in query.
+ * @param useAll    Mark the field is the leaf node in the tree.
+ * @param fields    Store the children of the field. It's safe to use
+ *                  name as the index because name is unique in every
+ *                  level.
+ * @param order     For leaf node, the order is used to memorize the
+ *                  location of the field in the new schema. For root
+ *                  node, it is used to memorize the number of leaf node.
+ */
+class RexNodeNestedField(
+    val name: String,
+    val index: Int,
+    val fieldType: RelDataType,
+    var useAll: Boolean,
+    var fields: JMap[String, RexNodeNestedField],
+    var order: Int) {
```

Comment:
The order is set after building the tree. It's not safe to set during the building. For example, if we have the expr "$0.loc" and "$0". It will first create the children of node "$0" and label it as the first left. However, when we meet the "$0", we need to first delete the children and itself as the first leaf.